src/library_pthread.js

```diff
@@ -164,9 +164,6 @@ var LibraryPThread = {
     freeThreadData: function(pthread) {
       if (!pthread) return;
       if (pthread.threadInfoStruct) {
-        var tlsMemory = {{{ makeGetValue('pthread.threadInfoStruct', C_STRUCTS.pthread.tsd, 'i32') }}};
-        {{{ makeSetValue('pthread.threadInfoStruct', C_STRUCTS.pthread.tsd, 0, 'i32') }}};
-        _free(tlsMemory);
```

Comment:
I just realized that this move to `_emscripten_thread_exit` could potentially cause issues with `pthread_cancel`, `pthread_kill` or detached threads, because they will not end up calling `pthread_exit` (and thus the thread-specific data associated with a thread is not being `free()`'d).

Reply:
Yes I noticed that too.  

I think for `pthread_cancel` we shouldn't have a problem because cancel now calls `_emscripten_thread_exit`: https://github.com/emscripten-core/emscripten/blob/58647ae5b0fa8af834871c26b3839f5bd84dd3d8/src/worker.js#L276

However, for pthread_kill I don't think there is much we can do.  I guess its reasonable that you can end up with leaks in that case. In fact I think we already have leaks because we don't call `__pthread_tsd_run_dtors` or anything that is registered via `__cxa_thread_atexit` (which means don't free the static TLS block: https://github.com/emscripten-core/emscripten/blob/58647ae5b0fa8af834871c26b3839f5bd84dd3d8/system/lib/pthread/emscripten_tls_init.c#L42).
Perhaps we can address this one case, but in general I think pthread_kill can always leak because the thread doesn't get any chance to shut itself down.  WDYT?

Reply:
Thanks for the clarification! I think we can't do much about `pthread_kill` then (since it doesn't run the thread-specific data dtors either, as you noticed). `pthread_cancel` does indeed no longer have this issue.

How about detached threads that are never joined and are shutdown during `PThread.terminateAllThreads()`? I think the proxied main thread (when linking with ` -sPROXY_TO_PTHREAD`) is such a case, but perhaps that is also negligible.

Reply:
I think `Thread.terminateAllThreads` is kind of the same as the `pthread_kill` case.. but we care even less about leaks in that case because the program is basically over.   

Its kind of shame we don't have a better way to force termination in the test itself.  In them musl code they use signals to bring down a thread which means it gets some time to tidy itself up a little, but we don't any way to interrupt threads on the web or in node.  

Perhaps we should just implement pthread_kill in terms of pthread_cancel and note that it won't work for threads that are busy-looping.   Either that or document that it leads to memory-leaks.

Reply:
I'm not sure about making `pthread_kill` non-compliant, I think documenting that it may leak memory is a better approach (or perhaps document that `pthread_cancel` is the preferred/safest way to terminate a thread).